### PR TITLE
Unpin Matplotlib and proj

### PR DESCRIPTION
**Pull request by stephenworsley**
Addresses #3759 and #3154.

It is expected that many tests may have to change to accommodate this. It should be demonstrated in each case that such a change is appropriate.

### CODE REVIEW

**Comment by stephenworsley** [comment]
The test `test_extent (iris.tests.unit.analysis.cartography.test_project.TestAll)` is failing due to a change in the numerical result. This may be due to improvements in the accuracy of Proj in the Robinson projection made, for example, here https://github.com/OSGeo/PROJ/pull/1292.

**Comment by trexfeathers** [comment]
Installing Cartopy now installs gdal ([see changes to the Cartopy recipe](https://github.com/conda-forge/cartopy-feedstock/commit/ec659aafe518a5125bda5e1500de1734a43ab5ad)). Is it therefore appropriate to remove from [iris/requirements/extensions.txt](https://github.com/SciTools/iris/blob/37f41065d5b3817b77040f8d8b0d268ba939d588/requirements/extensions.txt#L9), or is it good to keep this as 'belt and braces'?

**Comment by pp-mo** [comment]
> Is it therefore appropriate to remove GDAL from iris/requirements/extensions.txt
Looking at what was done here, I now think it was a mistake to put it in that list in the first place (and [I did it](https://github.com/SciTools/iris/pull/2952)!) : The comment suggests that the 'extensions' should be things that **depend on Iris itself** (hence iris-grib).  But equally, it's not clear when this information would ever used : not by Travis, not by pip, so presumably it's just a hint to users.

In short, I would remove it + not worry.

**Comment by stephenworsley** [comment]
I have a hunch that this PR is the one which caused the GDAL related tests to fail https://github.com/OSGeo/gdal/pull/1185.
Unfortunately, it's not a small PR so it's a bit tricky to unpick what's going on here, but relevant to our tests are these changes https://github.com/OSGeo/gdal/pull/1185/files#diff-3610c5e11cf9b7bc8e8b33baa7f4ca70. Specifically the change of certain strings from "unnamed" to unknown.

**Comment by trexfeathers** [comment]
In light of SciTools/cartopy#1105, I've changed all tests that use `ax.coastlines()` to explicitly state the target resolution. This will not yet fix the relevant graphics tests as there are still some minor border differences to accept - will review those with @stephenworsley tomorrow.

This change has also caused a failure in [test_plot_custom_aggregation](https://github.com/SciTools/iris/blob/master/docs/iris/gallery_tests/test_plot_custom_aggregation.py#L21-L26) ([see the relevant Travis build](https://travis-ci.org/github/SciTools/iris/jobs/720728321#L1466-L1477)), but in this case I propose we should accept the new result as the correct one since it will henceforth demonstrate Cartopy's `AdaptiveScalar` functionality, which makes Iris simpler to use. In this situation would it be appropriate to also remove the old 'out-dated' image hash(es) as no longer acceptable?

**Comment by trexfeathers** [comment]
Given the discussion on #2952, we should probably have a wider discussion about the implications of Cartopy now bringing in GDAL as a core dependency.
* Size: I compared the size of a conda environment for Iris master (a4541ce5b482f4740c9b74d1022c0a6f2c9a22e7) and this PR (da2977ea67bc33490c75af8aa9fb45c47bb853f7)
  I think this is a serious cause for concern, particularly at a time when we are discussing _reducing_ Iris' install size:
  * `python requirements/gen_conda_requirements.py --groups test > [my file name]`
  * `conda create -n [my env name] --file [my file name]`
  * `du -sh [path to my conda environment]`
  * master: `626M`
    #3762: `1.1G`
* Constraints/pins: [the GDAL recipe](https://github.com/conda-forge/libgdal-feedstock/blob/master/recipe/meta.yaml) is now significantly less pinned than it when at the time @pp-mo raised #2952
* Possible other concerns - I need to have a thorough read of #2952 to see how the discussion relates to the current scenario

**Comment by trexfeathers** [comment]
### Graphics test failures

Accepted image hash changes (SciTools/test-iris-imagehash#35):
* Since we are pinning `cartopy>=0.18`, and this changes some behaviours, I have elected to set new image hashes as the **only** acceptable targets for:
  * efda068c7141a82632a550c974f24b8455a71ea3 Shows off the auto-scaling coastlines (SciTools/cartopy#1105) in the [custom aggregation gallery example](https://github.com/SciTools/iris/blob/efda068c7141a82632a550c974f24b8455a71ea3/docs/iris/gallery_code/general/plot_custom_aggregation.py)
  * 40500121731f96f6fac26f5e2090ed3e906690c4 Reflects the new grid-line spacing (SciTools/cartopy@2f5e568a3747bd38a44addf88c74b1ce7691ad4a)
* 41b3b2a3e9c7cc1cef55a5a8ec614088cda0b837 I have accepted some minor changes in color map ranges (barely perceptible) based on the fact this has happened with previous Matplotlib shifts (e.g. SciTools/test-iris-imagehash#26 - compare [before](https://scitools.github.io/test-iris-imagehash/images/v4/e85e36cb95a19999876d4d3694b06c78c7396329958434c2cecb6c6d69ce3b92.png) and [after](https://scitools.github.io/test-iris-imagehash/images/v4/e85e36cb95b199999765cd3694b06478c7396329958434c2cecb6c6d69ce1b92.png)).

Genuine failures:
* The failures seen for [TestPlotCoordinatesGiven.text_tx.4](https://travis-ci.org/github/SciTools/iris/jobs/721121881#L7506-L7522) and [TestQuickplotCoordinatesGiven.test_tx.3](https://travis-ci.org/github/SciTools/iris/jobs/721121881#L7591-L7605) appear to be genuine errors.
From what I can tell the wrong scalar number is being attached to the time axis; the wrong number can be seen being handled by Matplotlib ([e.g. within ticker.ScalarFormatter](https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/ticker.py#L639-L653)), even though the actual time values coming from [iris.tests.stock.realistic_4d](https://github.com/SciTools/iris/blob/efda068c7141a82632a550c974f24b8455a71ea3/lib/iris/tests/stock/__init__.py#L561-L655) have not changed.

**Comment by trexfeathers** [comment]
### "What's New" Entries
Before merging, will need to sort out ["What's New" entries using the new format](https://scitools-iris.readthedocs.io/en/latest/developers_guide/documenting/whats_new_contributions.html?highlight=whatsnew#contributing-a-what-s-new-entry) (committed to master since this PR branched). Note from the [Contribution categories](https://scitools-iris.readthedocs.io/en/latest/developers_guide/documenting/whats_new_contributions.html?highlight=whatsnew#contribution-categories) that this also applies to internal changes such as the testing changes I have made.

At a bare minimum I expect this will involve getting hold of [docs/iris/src/whatsnew/latest.rst](https://github.com/SciTools/iris/blob/master/docs/iris/src/whatsnew/latest.rst), but we will need to be sure to avoid merge conflicts and we may need to bring over other changes too. Can hopefully sort things out using minimal manual new commits, so as to avoid a big merge commit or rebase which will ruin the repo's commit history.

**Comment by trexfeathers** [comment]
All the graphics tests are now passing. @stephenworsley I'll leave you to continue investigating the other test failures. Let me know if you need a hand 🙂

**Comment by stephenworsley** [comment]
While I still haven't got to the root cause of why the behaviour has changed, I've looked through the documentation for the key 42113 (https://gdal.org/drivers/raster/gtiff.html#nodata-value) and there is nothing to indicate that storing a fill value for data without masked points is in any way improper.

**Comment by pp-mo** [comment]
> While I still haven't got to the root cause of why the behaviour has changed, I've looked through the documentation for the key 42113 (https://gdal.org/drivers/raster/gtiff.html#nodata-value) and there is nothing to indicate that storing a fill value for data without masked points is in any way improper.

:+1: I'm good with this solution.  
Whatever has changed, it's in gdal + probably more correct as-is than as-was, and in any case, no error in Iris.

**Comment by stephenworsley** [comment]
It looks like there is still something odd going on with test_cube. I got it working in my own environment so that the numbers were correct, but it looks like there's still something going on with the Travis environment causing there to be `inf` values.

**Comment by stephenworsley** [comment]
The missing values mentioned above may be related to this PR https://github.com/OSGeo/PROJ/pull/2151.

**Comment by trexfeathers** [comment]
> Given the discussion on #2952, we should probably have a wider discussion about the implications of Cartopy now bringing in GDAL as a core dependency.

Several devs have further discussed offline. GDAL offers a number of benefits, and currently would only be responsible for a _fraction_ of Iris' total install 'bloat', so we expect will be fine to include. We are already familiar with use cases that make regular use of GDAL/Fiona with Iris.

If we make progress with reducing Iris' install size ('Iris Slimline'), GDAL's inclusion may be of greater concern. If so, there may be a case to alter Cartopy's inclusion of GDAL in its Conda recipe: the changes to that recipe for 0.18 (conda-forge/cartopy-feedstock#88) no longer reflect the fact that GDAL is actually still an _optional_ dependency - it can still be seen to be optional in [cartopy/INSTALL](https://github.com/SciTools/cartopy/blob/3565f789bc81bf4036af90be37465e9bf7378435/INSTALL#L111-L113) and the absence of GDAL in [the "Minimum dependencies matrix entry of cartopy/.travis.yml](https://github.com/SciTools/cartopy/blob/3565f789bc81bf4036af90be37465e9bf7378435/.travis.yml#L5-L7) (instead only brought in with Fiona in the ["Latest everything" matrix entry](https://github.com/SciTools/cartopy/blob/3565f789bc81bf4036af90be37465e9bf7378435/.travis.yml#L8-L10)).

**Comment by trexfeathers** [comment]
... are we there?

**Comment by stephenworsley** [comment]
@trexfeathers close I think, just needs some reviewing now.

**Comment by trexfeathers** [comment]
OK, please can you confirm that you are happy with the changes I have made? We discussed almost all of this last week.

I will review the changes you have made and once we are both happy I will merge.

File: lib/iris/tests/results/unit/analysis/cartography/project/TestAll/cube.cml [hunk 1]
```diff
@@ -74,10 +68,10 @@
...
       </coord>
       <coord datadims="[2]">
-        <dimCoord id="3589f6ce" points="[-7762639.61572, -6037608.59, -4312577.56429,
-		-2587546.53857, -862515.512857, 862515.512857,
-		2587546.53857, 4312577.56429, 6037608.59,
-		7762639.61572]" shape="(10,)" standard_name="projection_y_coordinate" units="Unit('m')" value_type="float64"/>
+        <dimCoord id="3589f6ce" points="[-7762639.19859, -6037608.26557, -4312577.33255,
+		-2587546.39953, -862515.46651, 862515.46651,
+		2587546.39953, 4312577.33255, 6037608.26557,
+		7762639.19859]" shape="(10,)" standard_name="projection_y_coordinate" units="Unit('m')" value_type="float64"/>
```
Comment (trexfeathers) [inline lib/iris/tests/results/unit/analysis/cartography/project/TestAll/cube.cml:74]: Change due to improved accuracy of Robinson projection (OSGeo/PROJ#1292) 👍

File: lib/iris/tests/results/unit/analysis/cartography/project/TestAll/cube.cml [hunk 2]
```diff
@@ -9,24 +9,21 @@
...
-		[43.3491677352, 51.7189413318, 46.2669886567,
-		 ..., 47.2804916021, 51.4709777887,
-		 42.1572752986]]" shape="(10, 10)" standard_name="latitude" units="Unit('degrees')" value_type="float64">
+		[84.2615071705, 54.8444503701, 26.7572382165,
+		 ..., 30.1637424823, 58.6051855118,
+		 86.9541974993],
+		[inf, 62.6851802716, 39.629088106, ...,
+		 42.3249307728, 64.8680485778, inf],
+		[inf, inf, 46.2669891845, ..., 47.2804926757,
+		 inf, inf]]" shape="(10, 10)" standard_name="latitude" units="Unit('degrees')" value_type="float64">
```
Comment (trexfeathers) [inline lib/iris/tests/results/unit/analysis/cartography/project/TestAll/cube.cml:26]: New `inf` values are an improvement for points outside the Robinson projection - they replace spurious 'nonsense numbers' (https://github.com/OSGeo/PROJ/pull/2151). It is unlikely that this change has negative downstream consequences - @stephenworsley has confirmed the data at these points is already masked and that basic metrics (e.g. mean, stdev) upon regridding are unaffected 👍

File: lib/iris/tests/results/unit/analysis/cartography/project/TestAll/cube.cml [hunk 3]
```diff
@@ -40,23 +37,20 @@
...
-		[17.2456810934, 2.60343563417, -15.3083834763,
-		 ..., 14.270914103, -4.55555121054,
-		 -17.7002310265]]" shape="(10, 10)" standard_name="longitude" units="Unit('degrees')" value_type="float64">
+		[-58.34983089, -70.8813785605, -58.2031036304,
+		 ..., 60.1459645527, 72.0019781785,
+		 20.6956240803],
+		[inf, -35.1608514733, -42.5369291003, ...,
+		 43.052784475, 31.4703511243, inf],
+		[inf, inf, -15.3083916055, ..., 14.2709222711,
+		 inf, inf]]" shape="(10, 10)" standard_name="longitude" units="Unit('degrees')" value_type="float64">
```
Comment (trexfeathers) [inline lib/iris/tests/results/unit/analysis/cartography/project/TestAll/cube.cml:53]: New `inf` values are an improvement for points outside the Robinson projection - they replace spurious 'nonsense numbers' (https://github.com/OSGeo/PROJ/pull/2151). It is unlikely that this change has negative downstream consequences - @stephenworsley has confirmed the data at these points is already masked and that basic metrics (e.g. mean, stdev) upon regridding are unaffected 👍

File: lib/iris/tests/results/unit/analysis/cartography/project/TestAll/cube.data.0.json [hunk 4]
```diff
@@ -1 +1 @@
-{"std": 0.22496461010978716, "min": 287.27520751953125, "max": 288.09292602539062, "shape": [2, 3, 10, 10], "masked": true, "mean": 287.67258522727275}
\ No newline at end of file
+{"std": 0.22496461010978716, "min": 287.27520751953125, "max": 288.0929260253906, "shape": [2, 3, 10, 10], "masked": true, "mean": 287.67258522727275}
```
Comment (trexfeathers) [inline lib/iris/tests/results/unit/analysis/cartography/project/TestAll/cube.data.0.json]: Change is due to regeneration of target results, which has removed 14th decimal place (likely due to changes with numpy versions etcetera since the last time this was generated). Change in value is within an acceptable tolerance 👍

File: lib/iris/tests/unit/analysis/cartography/test_project.py [hunk 5]
```diff
@@ -142,8 +142,8 @@ def test_extent(self):
             [
                 -17005833.33052523,
                 17005833.33052523,
-                -8625155.12857459,
-                8625155.12857459,
+                -8625154.6651,
+                8625154.6651,
```
Comment (trexfeathers) [inline lib/iris/tests/unit/analysis/cartography/test_project.py:146]: Change due to improved accuracy of Robinson projection (OSGeo/PROJ#1292) 👍

File: lib/iris/analysis/cartography.py [hunk 6]
```diff
@@ -599,6 +599,12 @@ def project(cube, target_proj, nx=None, ny=None):
         on the statistics of the data e.g. the mean and standard deviation
         will not be preserved.
 
+    .. warning::
+
+        If the target projection is non-rectangular, e.g. Robinson, the target
+        grid may include points outside the boundary of the projection. The
+        latitude/longitude of such points may be unpredictable.
+
```
Comment (trexfeathers) [inline lib/iris/analysis/cartography.py:607]: Such a warning should have already been present, but has become more apparent with the work done on this PR with the Robinson projection 👍

File: lib/iris/tests/experimental/test_raster.py [hunk 7]
```diff
@@ -77,6 +77,10 @@ def _check_tiff_export(self, masked, inverted=False):
             339,
             33550,
             33922,
+            42113,
+            # Don't add a check entry for this, as coding changed between gdal
+            # version 1 and 2.
+            # tif_header_entries[42113] = (u'1e+20',)
```
Comment (trexfeathers) [inline lib/iris/tests/experimental/test_raster.py:83]: See https://github.com/SciTools/iris/pull/3762#issuecomment-684913233

File: lib/iris/tests/unit/experimental/raster/test_export_geotiff.py [hunk 8]
```diff
@@ -124,9 +124,10 @@ def test_sphere(self):
             # String has embedded floating point values,
             # Test with values to N decimal places, using a regular expression.
             re_pattern = (
-                r'GEOGCS\["unnamed ellipse",DATUM\["unknown",'
-                r'SPHEROID\["unnamed",637....,0\]\],PRIMEM\["Greenwich",0\],'
-                r'UNIT\["degree",0.01745[0-9]*\]\]'
+                r'GEOGCS\["unknown",DATUM\["unknown",'
+                r'SPHEROID\["unknown",637....,0\]\],PRIMEM\["Greenwich",0\],'
+                r'UNIT\["degree",0.01745[0-9]*,AUTHORITY\["EPSG","9122"\]\],'
+                r'AXIS\["Latitude",NORTH\],AXIS\["Longitude",EAST\]\]'
```
Comment (trexfeathers) [inline lib/iris/tests/unit/experimental/raster/test_export_geotiff.py:130]: Changes/improvements in PROJ's WKT strings (OSGeo/gdal#1185) 👍

File: lib/iris/tests/unit/experimental/raster/test_export_geotiff.py [hunk 9]
```diff
@@ -145,9 +146,11 @@ def test_ellipsoid(self):
...
             # Test with values to N decimal places, using a regular expression.
             re_pattern = (
-                r'GEOGCS\["unnamed ellipse",DATUM\["unknown",'
-                r'SPHEROID\["unnamed",637....,375.117[0-9]*\]\],'
-                r'PRIMEM\["Greenwich",0\],UNIT\["degree",0.01745[0-9]*\]\]'
+                r'GEOGCS\["unknown",DATUM\["unknown",'
+                r'SPHEROID\["unknown",637....,375.117[0-9]*\]\],'
+                r'PRIMEM\["Greenwich",0\],UNIT\["degree",0.01745[0-9]*,'
+                r'AUTHORITY\["EPSG","9122"\]\],AXIS\["Latitude",NORTH\],'
+                r'AXIS\["Longitude",EAST\]\]'
```
Comment (trexfeathers) [inline lib/iris/tests/unit/experimental/raster/test_export_geotiff.py:153]: Changes/improvements in PROJ's WKT strings (OSGeo/gdal#1185) 👍

**Review by trexfeathers — COMMENTED** [review]
This review is a text recording of a voice conversation during pair programming, where @stephenworsley explained the changes that they have made in the PR.

File: requirements/extensions.txt [hunk 10]
```diff
@@ -6,4 +6,3 @@
 # first.
 
 iris-grib  #conda:
-gdal
```
Comment (trexfeathers) [inline requirements/extensions.txt:9]: See https://github.com/SciTools/iris/pull/3762#issuecomment-679058782

File: lib/iris/plot.py [hunk 11]
```diff
@@ -441,6 +441,11 @@ def _draw_2d_from_bounds(draw_method_name, cube, *args, **kwargs):
                     values = np.arange(data.shape[data_dim] + 1) - 0.5
                 else:
                     values = coord.contiguous_bounds()
+                    values = _fixup_dates(coord, values)
+                    if values.dtype == np.dtype(object) and isinstance(
+                        values[0], datetime.datetime
+                    ):
+                        values = mpl_dates.date2num(values)
```
Comment (stephenworsley) [inline lib/iris/plot.py:448]: The use of _fixup_dates and date2num is now more consistent with the rest of the code.

File: lib/iris/quickplot.py [hunk 12]
```diff
@@ -46,6 +47,13 @@ def _title(cube_or_coord, with_units):
 
             if _use_symbol(units):
                 units = units.symbol
+            if units.is_time_reference():
+                # iris.plot uses matplotlib.dates.date2num, which is fixed to the below unit.
+                if mpl_version >= "3.3":
+                    days_since = "1970-01-01"
+                else:
+                    days_since = "0001-01-01"
+                units = "days since {}".format(days_since)
```
Comment (stephenworsley) [inline lib/iris/quickplot.py:56]: The change in dates with 3.3 is due to https://github.com/matplotlib/matplotlib/pull/15008.

File: lib/iris/tests/results/imagerepo.json [hunk 13]
```diff
@@ -6,7 +6,8 @@
     "gallery_tests.test_plot_COP_maps.TestCOPMaps.test_plot_cop_maps.0": [
         "https://scitools.github.io/test-iris-imagehash/images/v4/ea9138db95668524913e6ac168997e85957e917e876396b96a81b5ce3c496935.png",
         "https://scitools.github.io/test-iris-imagehash/images/v4/ea9130db95668524913c6ac178995b0d956e917ec76396b96a853dcf94696935.png",
-        "https://scitools.github.io/test-iris-imagehash/images/v4/ea9130db95668524913e6ac168991f0d956e917ec76396b96a853dcf94796931.png"
+        "https://scitools.github.io/test-iris-imagehash/images/v4/ea9130db95668524913e6ac168991f0d956e917ec76396b96a853dcf94796931.png",
```
Comment (stephenworsley) [inline lib/iris/tests/results/imagerepo.json:9]: Changes in this file are expounded on in this comment: https://github.com/SciTools/iris/pull/3762#issuecomment-680249000. I've been over these changes with @trexfeathers and they all seem appropriate.

**Review by stephenworsley — COMMENTED** [review]
The changes due to graphical tests look good to me.

**Comment by trexfeathers** [comment]
@stephenworsley and I have finished pair programming and pair reviewing this PR. @pp-mo we'll give you some time to comment/review if you wish, but we consider this good to go 💯

**Comment by pp-mo** [comment]
>  @pp-mo we'll give you some time

Ok I've not considered everything here at all, only details, but I'll take a little look + see if anything sticks out...

**Comment by pp-mo** [comment]
Just to note from [previous comment](https://github.com/SciTools/iris/pull/3762#issuecomment-679271036) ...
> In this situation would it be appropriate to also remove the old 'out-dated' image hash(es) as no longer acceptable?

In short, probably *no*.  This would prevent running tests on an older checkout, and we generally prefer to leave it working.

**Comment by pp-mo** [comment]
A lot of very careful work here ! 💐 
Gets my vote.  @trexfeathers , when ready : hit that button :+1:
( you deserve it ! )

**Comment by trexfeathers** [comment]
> However, we definitely don't want to remove the images from test-iris-imagehash : That would prevent running the tests on an older checkout, and we would generally prefer to keep that working, if possible (though dependency changes may still mess it up).

I agree and accordingly I haven't done anything with [SciTools/test-iris-imagehash](https://github.com/SciTools/test-iris-imagehash).

**Comment by xylar** [comment]
I just noticed this discussion.  I'm sorry me adding `gdal` as a dependency of `cartopy` caused you trouble.  You are right in https://github.com/SciTools/iris/pull/3762#issuecomment-685573007 that it should have been kept optional.  I'm removing it as a required dependency in the next build: https://github.com/conda-forge/cartopy-feedstock/pull/94 but wanted to check in to make sure this isn't going to cause additional trouble.

**Comment by xylar** [comment]
Sorry, I spoke too soon -- the build seems to fail without it.

**Comment by trexfeathers** [comment]
> but wanted to check in to make sure this isn't going to cause additional trouble.

@xylar luckily we're currently doing some work that should flush out any possible recipe issues, so we should be able to let you know in the next couple of days. Don't think we're expecting any problems though 🙂

**Comment by xylar** [comment]
Okay, I'm going to go ahead an merge https://github.com/conda-forge/cartopy-feedstock/pull/94 unless you would prefer I hold off.
